### ISSUE TITLE
Please help. How to handle multiple requests by different users which one is which

### ISSUE THREAD

**Issue by sniperwolfpk5** [elsa-workflows/elsa-core#261]
Hello @sfmskywalker,

I do not want to use **signal activity**.

I have implemented Employee leave workflow. Where multiple employee can submit leave request to their managers. 

Let me explain through example. when **employee1** generate request which is now waiting for approval. The 2nd **employee2** generate request. Every thing is fine but when any manager perform action on leave request all other requests are moved to **Fault State**. I am setting correlation Id and also checking for correct correlate Id. I can't handle multiple requests. I have attached workflow with this post.

I will make your life easy. Just import attached workflow and perform below employee actions using Postman.

**Employee 1 Leave Request**
Post Below Json data using Postman
Post URI: http://localhost:5683/leaverequest
```
{
     "WorkflowId":"Leave_1",
    "RecordId": "1",
    "EntityName":"EmployeeLeave",
    "Applicant": {"Name": "Employee1", "Email": "qamar@abc.com"},
    "Sender": {" Name": "Employee1", "Email": "qamar@abc.com"},
    "Recipient": {"Name":"manager", "Email":"manager@abc.com"},
    "Action": "Initiated",
    "WorkflowStatus": "Start",
    "WorkflowDescription": "Leave Application submitted by ${Request.Sender.Name}",
    "Remarks": "Urgent work at home"
}
```
**WorkflowId**: is used to set correlation id
**RecordId**: Employee leave record id
**Action**:  user action performed possible actions (Initiated, Rejected, Forwarded, Withdrawn, 
Deleted, Approved etc.)
 **WorkflowStatus**: the state of workflow which is used by our application possible states (Start, InProgress, Correction, Denied, Canceled, Complete etc.)
**Employee 2 Leave Request**

Post URI: http://localhost:5683/leaverequest
```
{
	"WorkflowId":"Leave_2",
    "RecordId": "2",
    "EntityName":"EmployeeLeave",
    "Applicant": {"Name": "Employee2", "Email": "employee2@abc.com"},
    "Sender": {" Name": "Employee2", "Email": "employee2@abc.com"},
    "Recipient": {"Name":"manager", "Email":"manager@abc.com"},
    "Action": "Initiated",
    "WorkflowStatus": "Start",
    "WorkflowDescription": "Leave Application submitted by Employee2",
    "Remarks": "Urgent work at home"
}
```
No issues till now but when manager perform action all other requests are moved to Fault state. Lets do it

**Manager Action**
When manager perform action on any request
Post action request on the following Uri
**Note action URL is change**
Post URI: http://localhost:5683/leaveaction
```
{
	"WorkflowId":"Leave_2",
    "RecordId": "2",
    "EntityName":"EmployeeLeave",
    "Applicant": {"Name": "Employee2", "Email": "employee2@abc.com"},
    "Sender": {" Name": "manager", "Email": "manager@abc.com"},
    "Recipient": {"Name":"Employee2", "Email":"employee2@abc.com"},
    "Action": "Approved",
    "WorkflowStatus": "Complete",
    "WorkflowDescription": "Leave Application submitted by Employee2",
    "Remarks": "Urgent work at home"
}
```
Due to manager action all other requests are moved to Fault state. Only one request that is performed by Manager is fulfilled others are in Fault State. 

I do not know how to use UserTask activity. I tried to use Trigger activity but without any luck.

Please help how to use Httprequest with user actions with multiple employee requests. 

I highly appreciate your help

[workflow.(14).json.txt](https://github.com/elsa-workflows/elsa-core/files/4316431/workflow.14.json.txt)

**Comment by sniperwolfpk5** [comment]
@sfmskywalker  I get the following error  of the faulted state requests

`Response has already started`

**Comment by sfmskywalker** [comment]
Hi @sniperwolfpk5, thanks for the detailed description, it clarifies a lot! I guess it's needless to say that your workflow is a great example that covers many variations of a workflow where multiple users / workflow instances are involved and where **correlation** is key. I wish to document this in the form of a guide on the documentation site at some point.

Until I have time to do that (likely not before the release of Elsa 2, just to set expectations straight), here are a few pointers that might help.

The key to making this work is to use correlation. One way this could work in your scenario is by correlating your **leave request** workflow instances with your leave record ID that you post to the workflow.

Then, whenever the manager posts an action to /leaveaction, you must somehow also specify the correlation ID value so that Elsa only resumes the expected workflow instance.

The questions of course are:

1. How do you correlate the workflow instance created with using /leaverequest endpoint with the record ID
2. How do you specify the record ID as the correlation ID when triggering the workflow using the /leaveaction endpoint.

The first question is easy: simply use the `Correlate` activity. This activity expects a single workflow expression that evaluates to the value to be used as the correlation ID. In your case, it would be something like `{{ Content.RecordId }}` using Liquid syntax, or `(Content.RecordId)` using JS syntax (notice that the parentheses are required due to a weird quirk in the Jint interpreter when using simple expressions like this).

Now that the workflow instance is correlated to a known value (the record ID), let's tackle the second question: how to specify a record ID as a correlation ID.

The short answer: the HTTP middleware currently used doesn't provide an option to specify a correlation ID, so that's useless at the moment. However, it's a very important omission that we need to correct. Until that's fixed,one way to work around this limitation is to implement a controller yourself, which simply takes your JSON payload, read its Record ID value, and use the Elsa library to trigger the correct workflow yourself. Such code would look something like this (pseudo code):

```csharp
public LeaveController : ControllerBase
{
   public LeaveController(IWorkflowInstanceStore store, IWorkflowInvoker workflowInvoker)
   {
      _store = store;
      _workflowInvoker = workflowInvoker;
   }

   public async Task<IActionResult> Post(LeaveActionModel model)
   {
      var recordId = model.RecordId;

      // Get all workflow instances blocked by ReceiveHttpRequest activity and correlated with the record ID (probably in most typical scenarios this would yield just one workflow, but it's good practice to be flexible for more advanced use cases).
      var suspendedWorkflows = await _workflowInstanceStore.ListByBlockingActivityAsync<ReceiveHttpRequest>(correlationId: recordId);

      // For each found workflow instance, resume it.
      foreach (var (workflowInstance, activity) in suspendedWorkflows)
      {
          await _workflowInvoker.ResumeAsync(
              workflowInstance,
             Variables.Empty,
             new[] { activity.Id });
      }
   }
}
```

The above should work, but as I said I will fix the HTTP middleware so that you can simply provide the correlation ID value via a header, e.g. `x-correlation-id=2`.

Hope this helps! If not, please do not hesitate to let me know.

**Comment by sniperwolfpk5** [comment]
Hi @sfmskywalker ,

Thanks for the reply. Appreciate your detail answer. I have already fixed my workflow. I will share my steps how I fixed it to help others. 
As I am already setting correlation Id ("WorkflowId":"Leave_2") as mentioned in original post. It will make workflow instance unique. In `/leaverequest ` I set **correlation Id** and during `/leaveaction` I post correlation Id along with other data and check whether executing workflow instance is same. If request is same then normal actions are taken. If correlation id is not same then connect it to ReceiveHttpRequest (/leaveaction)

Long story short I removed `WriteHttpResponse` and Connected **Check Correlation Id** activity to **ReceiveHttpRequest (/leaveaction)** activity. Because `WriteHttpResponse` throws exception and when it throws exception it changes the state of executing workflow instance to Fault. As Elsa will loop through all blocking instances of leave workflow so all the other instances goes to Fault state only the one which has matching correlation id is executed successfully. Connecting **Check Correlation Id** activity to directly ReceiveHttpRequest (/leaveaction) will prevent exception, which will prevent other instances to move to Fault state. What I did is in below image

![updated workflow](https://user-images.githubusercontent.com/11886702/76588627-4251f580-6509-11ea-8327-bc38dddf2293.png)

Elsa has powerful feature to use it as API. I cannot use Elsa directly in my application as we are using Asp.net MVC 5. 

Attached is a fixed workflow which will work with multiple user scenario 
[workflow.(16).json.txt](https://github.com/elsa-workflows/elsa-core/files/4327814/workflow.16.json.txt)

**Comment by sfmskywalker** [comment]
I'm pleased to see you were able to work around it! Indeed, when a request comes in, *all* instances execute. Depending on your scenario, this may or may not be desirable. I like the way you solved it, but I also recognize that this solution may not be desirable in all cases. So I'd like to keep this issue open as a reminder for me to add support for executing single workflow instances filtered by their correlation ID as described earlier.

Thanks for explaining your work around in detail, it will help others understand.

**Comment by sniperwolfpk5** [comment]
Hi @sfmskywalker,

Yes you are right it will not work on all cases. Also it is not efficient way because it loops through all instances. A correlationId or workflow instance Id would be ideal with HTTP request which will execute only instance based on workflow instance id or correlation id.

**Comment by sfmskywalker** [comment]
@sniperwolfpk5 I'm closing this issue because I think we've come to a good conclusion and created a [separate issue](https://github.com/elsa-workflows/elsa-core/issues/262) to track development of it. Feel free to reopen this one if there's anything we need to follow up on. Thanks!
